File: modules/widgetkit/android/src/main/java/expo/modules/widgetkit/WidgetKitModule.kt
package expo.modules.widgetkit

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import com.google.gson.Gson
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import androidx.core.content.edit

class WidgetKitModule : Module() {
    companion object {
        const val groupName = "group.com.revcel.mobile"
        const val instancesKey = "revcel::connections"
        const val isSubscribedKey = "revcel::subscribed"
    }

    private fun prefs() = appContext.reactContext?.getSharedPreferences(groupName, Context.MODE_PRIVATE)

    private fun getConnections(): List<Connection> {
        val rawConnections = prefs()?.getString(instancesKey, "[]")

        return rawConnections?.let {
            runCatching { Gson().fromJson(it, Array<Connection>::class.java).toList() }.getOrNull()
        } ?: emptyList()
    }

    /** Writes the list and notifies the widgets only when the payload changed. */
    private fun saveConnections(connections: List<Connection>) {
        val prefs = prefs() ?: return
        val serialized = Gson().toJson(connections)

        if (prefs.getString(instancesKey, null) == serialized) return

        prefs.edit {
            putString(instancesKey, serialized)
            apply()
        }

        notifyAllWidgets()
    }

    private fun notifyAllWidgets() {
        val context = appContext.reactContext ?: return

        // implicit broadcasts are dropped since API 26; scoping to our package delivers it to the
        // manifest receivers of every widget
        val intent = Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE).apply {
            setPackage(context.packageName)
        }

        context.sendBroadcast(intent)
    }

    override fun definition() = ModuleDefinition {
        Name("RevcelWidgetKit")

        Function("setIsSubscribed") { isSubscribed: Boolean ->
            prefs()?.let { prefs ->
                if (prefs.contains(isSubscribedKey) && prefs.getBoolean(isSubscribedKey, false) == isSubscribed) {
                    return@let
                }

                prefs.edit {
                    putBoolean(isSubscribedKey, isSubscribed)
                    apply()
                }

                notifyAllWidgets()
            }
        }

        Function("addConnection") { connection: Connection ->
            val connections = this@WidgetKitModule.getConnections().toMutableList()

            val index = connections.indexOfFirst { it.id == connection.id }

            if (index != -1) {
                connections[index] = connection
            } else {
                connections.add(connection)
            }

            saveConnections(connections)
        }

        Function("removeConnection") { id: String ->
            saveConnections(this@WidgetKitModule.getConnections().filter { it.id != id })
        }

        Function("clearAllConnections") {
            prefs()?.let { prefs ->
                // only the connections: clearing everything also dropped the subscription flag
                prefs.edit() {
                    remove(instancesKey)
                    apply()
                }

                notifyAllWidgets()
            }
        }
    }
}

File: modules/widgetkit/ios/WidgetKitModule.swift
import ExpoModulesCore
import WidgetKit

public class WidgetKitModule: Module {
    let _groupName: String = "group.com.revcel.mobile"
    let _connectionsKey: String = "revcel::connections"
    let _isSubscribedKey: String = "revcel::subscribed"
    let _projectEntitiesKey: String = "revcel::projectEntities"
    
    private func getConnections() -> [Connection] {
        guard let sharedDefaults = UserDefaults(suiteName: _groupName),
              let rawExistingConnections = sharedDefaults.data(forKey: _connectionsKey) else { return [] }
        
        return (try? JSONDecoder().decode([Connection].self, from: rawExistingConnections)) ?? []
    }
    
    private func reloadWidgets() {
        if #available(iOS 16.0, *) {
            WidgetCenter.shared.invalidateConfigurationRecommendations()
        }
        
        WidgetCenter.shared.reloadAllTimelines()
    }

    /// Writes the list and reloads the widgets only when the payload changed.
    private func saveConnections(_ connections: [Connection]) throws {
        guard let sharedDefaults = UserDefaults(suiteName: _groupName) else { return }

        let encodedConnections = try JSONEncoder().encode(connections)

        if sharedDefaults.data(forKey: _connectionsKey) == encodedConnections {
            return
        }

        sharedDefaults.set(encodedConnections, forKey: _connectionsKey)
        self.reloadWidgets()
    }
    
    public func definition() -> ModuleDefinition {
        Name("RevcelWidgetKit")
        
        Function("setIsSubscribed") { (isSubscribed: Bool) -> Void in
            guard let sharedDefaults = UserDefaults(suiteName: _groupName) else { return }

            if sharedDefaults.object(forKey: _isSubscribedKey) != nil,
               sharedDefaults.bool(forKey: _isSubscribedKey) == isSubscribed {
                return
            }
            
            sharedDefaults.set(isSubscribed, forKey: _isSubscribedKey)
            
            self.reloadWidgets()
        }
        
        Function("addConnection") { (connection: Connection) -> Void in
            var connections = self.getConnections()

            if let index = connections.firstIndex(where: { $0.id == connection.id }) {
                connections[index] = connection
            } else {
                connections.append(connection)
            }

            try? self.saveConnections(connections)
        }
        
        Function("removeConnection") { (id: String) in
            try? self.saveConnections(self.getConnections().filter { $0.id != id })
        }
        
        Function("clearAllConnections") {
            guard let sharedDefaults = UserDefaults(suiteName: _groupName) else {
                return
            }
            
            // only the connections: wiping the whole suite also dropped the subscription flag
            sharedDefaults.removeObject(forKey: _connectionsKey)
            sharedDefaults.removeObject(forKey: _projectEntitiesKey)
            
            self.reloadWidgets()
        }
    }
}
